### PR TITLE
Compilation of Js Files

### PR DESCRIPTION
**Pull request by sheetalkamat**
Compiler treats javascript files like declaration files and does not emit any output corresponding to them. Here are the list of work this PR supports:
- [x] Allow javascript files to be compiled on command line or through references/imports
- [x] tsconfig.json can take javascript files as file names to be compiled
- [x] If filenames are not specified in tsconfig.json, javascript files in the folder and not part of compilation
- [x] compiler options now have a flag named "consumeJsFiles" which when true will pick up javascript files in the folder for compilation if tscconfig.json doesn't supply file name list
- [x] Moved reporting of usage of typescript syntax in javascript file from services to compiler
- [x] Reports error if output is overwriting any of the input files
- [x] Reports error if output is not unique location and multiple input files overwrite same output file
- [x] It is error to reference .js file in a typescript file when generating declaration file.
- [x] .ts, .tsx and .d.ts files take precedence over .js files when creating list of files for compilation
- [x] Project runner options cleanup
- [x] Project runner to accept tsconfig.json tests
- [X] Instead of consumeJsFiles option, support jsExtensions that allows user to specify comma separated list of extensions that are to be treated like javascript files
- [ ] Bundle js files into --out and copy into --outDir

### CODE REVIEW

File: src/compiler/commandLineParser.ts [hunk 1]
```diff
@@ -376,6 +375,63 @@ namespace ts {
         }
     }
 
+    function parseMultiValueStringArray(s: string, existingValue: string[]) {
```
Comment (vladima) [inline src/compiler/commandLineParser.ts]: as I can see this function parses comma separated non-quoted strings with assumption that value itself should not have commas inside, i.e. `"a,b", "c"`. Can you document this behavior (it is not completely clear otherwise by looking at `string[]` as an argument type)

File: src/compiler/commandLineParser.ts [hunk 2]
```diff
@@ -405,22 +461,57 @@ namespace ts {
...
+                }
+                else {
+                    errors.push(createCompilerDiagnostic((<CommandLineOptionOfCustomType>opt).error));
+                    jsonValue = 0;
+                }
+            }
+        }
+        // Check if the value asked was string[] and value provided was not string[]
+        else if (expectedType !== "string[]" ||
+            typeof jsonValue !== "object" ||
```
Comment (vladima) [inline src/compiler/commandLineParser.ts]: `!Array.isArray(jsonValue)` ?

File: src/compiler/commandLineParser.ts [hunk 3]
```diff
@@ -474,23 +547,31 @@ namespace ts {
...
-                        }
-                    }
-                    else if (fileExtensionIs(name, ".ts")) {
-                        if (!contains(sysFiles, name + "x")) {
-                            fileNames.push(name);
+                let extensionsToRead = getSupportedExtensions(options);
+                for (let extensionsIndex = 0; extensionsIndex < extensionsToRead.length; extensionsIndex++) {
+                    let extension = extensionsToRead[extensionsIndex];
+                    let sysFiles = host.readDirectory(basePath, extension, exclude);
+                    for (let i = 0; i < sysFiles.length; i++) {
```
Comment (vladima) [inline src/compiler/commandLineParser.ts]: for-of` ?

File: src/compiler/core.ts [hunk 4]
```diff
@@ -714,20 +714,20 @@ namespace ts {
...
-        return pathLen > extLen && path.substr(pathLen - extLen, extLen) === extension;
+        let extLen = extension.length + 1;
+        return pathLen > extLen && path.substr(pathLen - extLen, extLen) === "." + extension;
     }
 
     /**
      *  List of supported extensions in order of file resolution precedence.
      */
-    export const supportedExtensions = [".ts", ".tsx", ".d.ts"];
+    export const supportedTypeScriptExtensions = ["ts", "tsx", "d.ts"];
```
Comment (vladima) [inline src/compiler/core.ts]: can you please elaborate why this change (dropping dots) is necessary?

File: src/compiler/commandLineParser.ts [hunk 5]
```diff
@@ -376,6 +375,63 @@ namespace ts {
         }
     }
 
+    function parseMultiValueStringArray(s: string, existingValue: string[]) {
+        let value: string[] = existingValue || [];
+        let hasError: boolean;
```
Comment (vladima) [inline src/compiler/commandLineParser.ts]: as a matter of preference, can it be `hasError = false`. I'm not a big fan of returning undefined as primitive if `s` is falsy (though it is totally legit)

File: src/compiler/core.ts [hunk 6]
```diff
@@ -714,20 +714,20 @@ namespace ts {
...
 
-    const extensionsToRemove = [".d.ts", ".ts", ".js", ".tsx", ".jsx"];
+    const extensionsToRemove = ["d.ts", "ts", "js", "tsx", "jsx"];
```
Comment (mhegazy) [inline src/compiler/core.ts]: we should not be using this any more, since users can specify additional jsExtenstions, this will fail unexpectedly.

File: src/compiler/program.ts [hunk 7]
```diff
@@ -14,10 +14,10 @@ namespace ts {
     
     export const version = "1.7.0";
 
-    export function findConfigFile(searchPath: string): string {
+    export function findConfigFile(searchPath: string, moduleResolutionHost: ModuleResolutionHost): string {
```
Comment (mhegazy) [inline src/compiler/program.ts]: this will be a breaking changes

Comment (mhegazy) [inline src/compiler/program.ts]: similar to `readConfigFile` we should just pass in fileExists method instead of the host.

File: src/harness/projectsRunner.ts [hunk 8]
```diff
@@ -203,23 +181,105 @@ class ProjectRunner extends RunnerBase {
...
+                };
+                let errors: ts.Diagnostic[] = [];
+                // Set the values specified using json
+                let optionNameMap: ts.Map<ts.CommandLineOption> = {};
+                ts.forEach(ts.optionDeclarations, option => {
+                    optionNameMap[option.name] = option;
+                });
+                for (let name in testCase) {
+                    if (name !== "mapRoot" && name !== "sourceRoot" && ts.hasProperty(optionNameMap, name)) {
+                        let option = optionNameMap[name];
```
Comment (mhegazy) [inline src/harness/projectsRunner.ts]: why not use the one in harness?

File: src/compiler/commandLineParser.ts [hunk 9]
```diff
@@ -474,23 +547,31 @@ namespace ts {
...
-                        let baseName = name.substr(0, name.length - ".d.ts".length);
-                        if (!contains(sysFiles, baseName + ".tsx") && !contains(sysFiles, baseName + ".ts")) {
-                            fileNames.push(name);
-                        }
-                    }
-                    else if (fileExtensionIs(name, ".ts")) {
-                        if (!contains(sysFiles, name + "x")) {
-                            fileNames.push(name);
+                let extensionsToRead = getSupportedExtensions(options);
+                for (let extensionsIndex = 0; extensionsIndex < extensionsToRead.length; extensionsIndex++) {
```
Comment (vladima) [inline src/compiler/commandLineParser.ts]: honestly logic below looks complicated.
Maybe it can be simplified to something easier to understand like:
```javascript
const extensionsByPriority = [.tsx, .ts, d.t.s] ...
let filesSeen = {};
let files = [];
for (let i = 0; i < extensionsByPriority.length; i++) {
    const ext = extensionsByPriority[i];
    let filesInDir = host.readDirectory(basePath, ext, exclude);
    // find all extensions in the list that
    // - longer than current extension
    // - ends with text of current extension
    // i.e. for ext='.ts' conflicting extension can be ['.d.ts']
    let conflictingExtensions = filter(extensionsByPriority, /*startIndex*/i + 1, e => e.length > ext.length && e.indexOf(ext, e.length - ext.length) !== -1);
    for (const fileName in filesInDir) {
        if (conflictingExtensions.some(e => fileExtensionIs(fileName, e))) {
            // file extension matches with some larger extension that was not checked yet
            // 'f.d.ts' will match '.ts' extension but really should be process later with '.d.ts' files
            continue;
        }        
        const baseName = fileName.substr(0, fileName.length - ext.length - 1);
        if (!hasProperty(filesSeen, baseName)) {
            filesSeen[baseName] = true;
            files.push(fileName);
        }
    }
}
```

File: src/harness/projectsRunner.ts (hunk 8 above)
Comment (sheetalkamat) [inline src/harness/projectsRunner.ts]: Because this is already in json format unlike strings in harness

File: src/compiler/commandLineParser.ts [hunk 10]
```diff
@@ -373,6 +372,68 @@ namespace ts {
     }
 
     /**
+     * Parses non quoted strings separated by comma e.g. "a,b" would result in string array ["a", "b"]
+     * @param s
+     * @param existingValue
+     */
+    function parseMultiValueStringArray(s: string, existingValue: string[]) {
```
Comment (RyanCavanaugh) [inline src/compiler/commandLineParser.ts]: What's the difference between this and the built-in `split` ?

Comment (sheetalkamat) [inline src/compiler/commandLineParser.ts]: I could use split but then I would need to iterate over result again to make sure the strings are non empty and are correct.

**Comment by sheetalkamat** [comment]
This is now handled in #5345
